### PR TITLE
Technical Challenge2:Auction algorithm

### PR DESCRIPTION
**Pull request by tobygabriella**
- Implemented algorithm to assign winners to each auction items
- Handled edge case for when users have the same score for same bidding: 
- Screenshot 2024-07-19 at 11 55 46 AM

- Another Bid result with multiple bidders: 
Screenshot 2024-07-19 at 12 12 33 PM

- Front end notifications for bid winners(most recent win): 
Screenshot 2024-07-19 at 3 12 06 PM

Screenshot 2024-07-19 at 3 12 57 PM

Screenshot 2024-07-19 at 3 13 29 PM

- Backend Logs for assigning winners: 
Screenshot 2024-07-19 at 3 11 31 PM

### CODE REVIEW

File: Backend/services/auctionSystem.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  calculateBidScore(bid) {
+    const maxBid = Math.max(
+      ...this.bids.filter((b) => b.itemId === bid.itemId).map((b) => b.amount)
+    );
+    const normalizedBid = maxBid > 0 ? bid.amount / maxBid : 0;
+    const userBidsForTheDay = this.bids.filter(
+      (b) => b.bidder.id === bid.bidder.id
+    ).length;
+    const normalizedBidsPerDay = Math.min(userBidsForTheDay / 10, 1);
+    const normalizedTime = (Date.now() / 1000 - bid.timestamp) / 86400;
```
Comment (conniewu3) [inline Backend/services/auctionSystem.js:20]: could we add some comments as to how you chose 10 and 1000 - bid.timestamp and 86400?

File: Backend/services/auctionSystem.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  getHighestScoringBid(bids) {
+    return bids.reduce((max, bid) => {
+      const scoreA = this.calculateBidScore(max);
+      const scoreB = this.calculateBidScore(bid);
+
+      if (scoreB !== scoreA) {
+        return scoreB > scoreA ? bid : max;
+      }
+      // Tie-breaking rules
+      if (bid.amount !== max.amount) {
```
Comment (conniewu3) [inline Backend/services/auctionSystem.js:104]: nice logic!

File: Backend/routes/listings.js [hunk 3]
```diff
@@ -137,7 +135,7 @@ router.get("/listings/:id", logActivity(), async (req, res) => {
   }
 });
 
-router.delete("/listings/:id", verifyToken, async (req, res) => {
+router.delete("/listings/:id", verifyToken, verifyToken, async (req, res) => {
```
Comment (conniewu3) [inline Backend/routes/listings.js:138]: did you mean to add logActivity instead of verifyToken twice here?

File: Backend/routes/listings.js [hunk 4]
```diff
@@ -9,12 +9,9 @@ const { sendNotification } = require("../services/notificationService");
 const NotificationType = require("../Enums/NotificationType.js");
 const logActivity = require("../middlewares/logActivity");
 const stripe = require("stripe")(process.env.STRIPE_SECRET_KEY);
-const { AuctionSystem } = require("../services/auctionSystem");
 
 const prisma = new PrismaClient();
-const auctionSystem = new AuctionSystem(prisma);
 const router = express.Router();
-router.use(verifyToken);
```
Comment (conniewu3) [inline Backend/routes/listings.js:15]: any reason why we remove this? might have been easier to just have router use verifyToken so we don't have to add it everywhere right?

File: Backend/routes/listings.js [hunk 5]
```diff
@@ -859,4 +862,46 @@ router.post(
...
 
+router.post("/run-daily-auctions", async (req, res) => {
+  const io = req.app.locals.io;
+  try {
+    const processedAuctions = await auctionSystem.processEndingAuctions();
+
+
+   // Get the assignments (winning bids) from the auction system
+    const assignments = auctionSystem.getAssignments();
+    console.log("the winning bids are", assignments);
```
Comment (conniewu3) [inline Backend/routes/listings.js]: remove this!

**Review by conniewu3 — COMMENTED** [review]
few questions and comments but otherwise looks great! nice job on this iteration of TC 2 :)

File: Backend/routes/listings.js (hunk 4 above)
Comment (tobygabriella) [reply Backend/routes/listings.js:15]: Yeah, it was interfering with my run auction route which did not require it. While I was not using the verify token in the file, it was still being affected by it. One of the solutions i tired was removing the global verify token from all files and it worked, so I decided to use it on individual routes instead.

File: Backend/routes/listings.js (hunk 3 above)
Comment (tobygabriella) [reply Backend/routes/listings.js:138]: yes thats right

**Pull request merged**